File: java/src/main/java/com/example/springsample/controller/support/AuthenticateSupport.java
package com.example.springsample.controller.support;

import com.example.springsample.model.User;
import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.ModelAttribute;

import java.util.Optional;

public interface AuthenticateSupport {

    @ModelAttribute
    default Optional<User> authenticate(@CookieValue(name = "sid", required = false) Optional<String> sessionId) {
        return sessionId.flatMap(this::findAccountBySessionId);
    }

    @ModelAttribute
    default User authenticate(@CookieValue(name = "sid") String sessionId) {
        return findAccountBySessionId(sessionId).get();
    }

    default Optional<User> findAccountBySessionId(String sessionId) {
        return Optional.ofNullable(sessionId)
                .filter(sid -> sid.equals("exists"))
                .map(sid -> new User("John Doe"));
    }

}

File: scala/src/main/scala/com/example/springsample/controller/support/AuthenticateSupport.scala
package com.example.springsample.controller.support

import com.example.springsample.model.User
import com.example.springsample.repository.UserRepository
import org.springframework.web.bind.annotation.{CookieValue, ModelAttribute}

trait AuthenticateSupport {

  val userRepository: UserRepository

  @ModelAttribute
  def allowGuest(@CookieValue(name = "sid", required = false) sessionId: String): Option[User] = {
    for {
      sid <- Option(sessionId)
      user <- userRepository.findBySessionId(sid)
    } yield user
  }
}
